test/Spec/BeneficiariesFile.hs

```diff
@@ -1,35 +1,156 @@
+  where
+    f inFile inConfig =
+      parseContent (defaultConfig {dropAmount = Nothing, assetClass = defaultConfig.assetClass <* guard inConfig})
+        . unwords
+        $ ["adfd87319bd09c9e3ea10b251ccb046f87c5440343157e348c3ac7bd"]
+          <> ["1000000"]
+          <> ["1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e.testToken" | inFile]
+
+prop_TokenAmountTruncation :: TokenAmount -> Positive Int -> Property
+prop_TokenAmountTruncation (TokenAmount amt) (Positive dp) =
+  let s = "adfd87319bd09c9e3ea10b251ccb046f87c5440343157e348c3ac7bd " <> show amt <> " 1d6445ddeda578117f393848e685128f1e78ad0c4e48129c5964dc2e.testToken"
       shouldTruncate = dp + base10Exponent amt < 0
-      didTruncate = isLeft $ parseContent (defaultConfig {decimalPlaces = toInteger dp}) (pack s)
+      didTruncate = isLeft $ parseContent (defaultConfig {decimalPlaces = toInteger dp, dropAmount = Nothing, assetClass = Nothing}) (pack s)
    in didTruncate === shouldTruncate
 
+prop_Parse :: Beneficiaries -> Property
+prop_Parse (Beneficiaries (bens, config)) = either (error . unpack) (const $ property True) $ parseContent config bens
+
+-- TODO Generate [Beneficiary] directly, don't depend on parsing
+prop_Print :: Beneficiaries -> Property
+prop_Print (Beneficiaries (bens, config)) = either (error . unpack) (const $ property True) $ parseContent config bens >>= prettyContent config
+
+prop_Roundtrip1 :: Beneficiaries -> Property
+prop_Roundtrip1 (Beneficiaries (bens, config)) =
+  -- We can't check for equality immediately from the incoming Text, as information can be lost during parsing.
+  -- Thus we initially parse and then print, and then do the roundtrip, as parse >=> print should be idempotent
+  let printed = either (error . unpack) id $ parseContent config bens >>= prettyContent config
+      roundTrip = parseContent config >=> prettyContent config
+   in roundTrip printed === Right printed
+
+-- TODO Generate [Beneficiary] directly
+prop_Roundtrip2 :: Beneficiaries -> Property
+prop_Roundtrip2 (Beneficiaries (bens, config)) =
+  let parsed = either (error . unpack) id $ parseContent config bens
+      roundTrip = prettyContent config >=> parseContent config
+   in roundTrip parsed === Right parsed
+
+newtype TokenAmount = TokenAmount {unTokenAmount :: Scientific}
+  deriving stock (Show)
+
+newtype Beneficiaries = Beneficiaries (Text, Config) deriving stock (Show)
+
+instance Arbitrary TokenAmount where
+  arbitrary = do
+    Positive (Large x) <- arbitrary @(Positive (Large Int))
+    Small scale <- arbitrary @(Small Int)
+    pure . TokenAmount $ normalize $ scientific (toInteger x) scale
+
+instance Arbitrary Beneficiaries where
+  arbitrary = do
+    usePK <- arbitrary @Bool
+    Positive dp <- arbitrary @(Positive Integer)
+
+    dropAmount' <- fmap unTokenAmount <$> arbitrary @(Maybe TokenAmount)
+    assetClass' <- (defaultConfig.assetClass <*) <$> arbitrary @(Maybe ())
```

Comment:
This feels weird
`Maybe ()` is just a boolean - could we shift the arbitrary to bool then build up what we need with the pure values?
even if it is something like
```hs
useAsset <- arbitrary @Bool
let assetClass' = if useAsset then Just defaultConfig.assetClass else Nothing
```
Theres probably some utils you can use to clean this up - like `boolToMaybe` or `ifThenElse`, the logistics of it don't really matter. I just feel like this conbination of operators and odd arbitrary instance look strange

Reply:
Oh! You can actually do 
```hs
assetClass' <- liftArbitrary $ pure defaultConfig.assetClass
```
But its worth noting that the liftArb for Maybe is a 3 to 1 of Just to Nothing
Unsure if this is desirable.

Reply:
liftArb wouldn't work directly on defaultConfig.assetClass as it's inside Maybe already, just went ahead and inlined the value of defaultConfig.assetClass and used liftArb; there's no reason it has to be the default value, it was just convenient.